File: src/main/scala/com/vc/hard/L514.scala
package com.vc.hard

object L514 {
  def findRotateSteps(ring: String, key: String): Int = {
    import scala.collection.mutable
    val map = new mutable.HashMap[String, mutable.HashMap[Int, Int]]()

    def find(ring: String, ch: Char): Int = {
      var i = ring.length - 1
      while(i >= 0) {
        if(ring(i) == ch) return i
        i -= 1
      }
      i
    }

    def solve(ring: String, index: Int): Int = {
      if(index == key.length) return 0
      val ch = key(index)
      if(map.contains(ring) && map(ring).contains(index)) return map(ring)(index)
      val forwardMove = ring.indexOf(ch)
      val backwardMove = find(ring, ch)
      val forwardTotal = 1 + forwardMove + solve(ring.substring(forwardMove) + ring.substring(0, forwardMove), index + 1)
      val backwardTotal = 1 + ring.length - backwardMove + solve(ring.substring(backwardMove) + ring.substring(0, backwardMove), index + 1)
      val res = Math.min(forwardTotal, backwardTotal)
      val ringMap = map.getOrElse(ring, new mutable.HashMap[Int, Int])
      ringMap.put(index, res)
      map.put(ring, ringMap)
      res
    }
    solve(ring, 0)
  }
}

File: src/main/java/com/vc/hard/L514.java
package com.vc.hard;

import java.util.Arrays;

class L514 {

    int[][] memo;
    public int findRotateSteps(String ring, String key) {
        this.memo = new int[ring.length() + 1][key.length() + 1];
        for(int i = 0; i < memo.length; i++) {
            Arrays.fill(memo[i], -1);
        }
        return helper(ring, key, 0, 0);
    }

    private int helper(String ring, String key, int ringIndex, int keyIndex) {
        if(keyIndex == key.length()) return 0;

        if(memo[ringIndex][keyIndex] != -1) return memo[ringIndex][keyIndex];

        char ch = key.charAt(keyIndex);
        int min = Integer.MAX_VALUE;
        for(int i = 0; i < ring.length(); i++) {
            if(ring.charAt(i) != ch) continue;

            int distance = Math.abs(ringIndex - i);
            int minDistance = 1 + Math.min(distance, ring.length() - distance);

            min = Math.min(min, minDistance + helper(ring, key, i, keyIndex + 1));
        }

        memo[ringIndex][keyIndex] = min;
        return min;
    }
}